# Trabalho Matemática Discreta:
**Integrantes: Marney Melo e Rafael Miranda**

# Inteiros

Veja também Rosen Seção 3.4, 3.5 (final) e 5.6 (final).

## Algoritmo de Divisão

**Teorema** *$\forall\ a,b \in \mathbb{Z}$, $\exists\ q,r  \in \mathbb{Z}$ tais que $a=q\cdot b+r$ com $0\leq r<b$. Aqui, $q$ é o quociente e $r$ é o resto.*

Uma prova formal deste fato bem conhecido pode ser obtida por indução em $|a|$.

- O comando `q , r = divmod(a, b)` em Python executará a operação correspondente.
- O comando `q = a // b` retornará o quociente, enquanto `r = a % b` retornará o resto. *ATENÇÃO:* o comando `q = a/b` sempre retornará um float, o que não é o que queremos neste contexto.

. . .

**Definição** Dizemos que $b$ divide $a$, ou que $b$ é um divisor de $a$, ou que $a$ é um múltiplo de $b$ se $\exists q \in \mathbb{Z}$ tal que $qb = a$. Notação: $b \mid a$.

Exemplos: $5 \mid 15; 6 \not\mid 15; 0 \mid 0; 1 \mid -1$

A seguinte expressão booleana em Python testa a divisibilidade: `(a%b==0)`. Se quiser empacotar em uma função, temos:

In [291]:
def DIV(b,a):
	return (a % b == 0)

Mas na prática manter a expressão original é mais fácil.

**<span style="color: red;"> [1.1] Teste de exemplos: </span>**

In [5]:
def div(b, a):
    if b == 0 and a == 0:
        print(b, '|', a)  # Por definição, 0 divide 0
    elif b == 0:
        print(b, '∤',a)  # 0 não divide nenhum número, exceto 0
    elif a % b == 0:
        print(b, '|', a)
    else:
        print(b, '∤',a)

test_cases = [(3, 15), (6, 15), (0, 0), (1, -1)]

for b, a in test_cases:
    div(b, a)

3 | 15
6 ∤ 15
0 | 0
1 | -1



**Propiedades**
$$
\begin{array}{ll}
i)\quad & c \mid b \wedge b \mid a ~\Rightarrow ~ c \mid a \\
ii)\quad & b \mid a \wedge b \mid a' ~\Rightarrow ~ b \mid a\mbox{+}a' \\
iii)\quad & b \mid 0 ~~\forall b \in \mathbb{Z} \\
iv)\quad & 1 \mid a ~~\forall a \in \mathbb{Z} \\
v) &0 \mid a \iff a=0 \\
\end{array}
$$

**<span style="color: red;"> [1.2] Teste estas propriedades </span>**

In [6]:
#Verifica se b divide a (retorna True/False).
def divide(b, a):
     if b == 0:
        return a == 0  # 0 só divide 0
     return a % b == 0  # Caso geral

**<span style="color: red;"> Função auxiliar para divisibilidade. </span>**

##### Propriedade (i) - Transitividade  
$ i)\quad c \mid b \land b \mid a \Rightarrow c \mid a  $  

In [7]:
exemplos_1 = [(2, 6, 12), (3, 9, 27), (7,14,84), (4, 8, 15)]

def transitividade():
    for c, b, a in exemplos_1:
        # Verifica se as premissas são verdadeiras (c|b E b|a)
        premissas_validas = divide(c, b) and divide(b, a)
        
        # Se as premissas forem válidas, verifica a conclusão (c|a)
        if premissas_validas:
            conclusao = divide(c, a)
            print(f"{c}|{b} e {b}|{a} ⇒ {c}|{a}? {conclusao} (Transitividade válida)")
        else:
            # Mostra qual premissa falhou
            if not divide(c, b):
                print(f"{c}∤{b} (Premissa c|b inválida) ⇒ Transitividade não aplicável")
            elif not divide(b, a):
                print(f"{b}∤{a} (Premissa b|a inválida) ⇒ Transitividade não aplicável")
transitividade()

2|6 e 6|12 ⇒ 2|12? True (Transitividade válida)
3|9 e 9|27 ⇒ 3|27? True (Transitividade válida)
7|14 e 14|84 ⇒ 7|84? True (Transitividade válida)
8∤15 (Premissa b|a inválida) ⇒ Transitividade não aplicável


##### Propriedade (ii) - Linearidade:
##### $ii) b \mid a \land b \mid a' \Rightarrow b \mid a+a'$ 

In [8]:
exemplos_2 = [(2,4,10), (3,9,30), (5,15,85), (9,36,49)]

def lineariedade():
    for c, b, a in exemplos_2:
        # Verifica se as premissas são verdadeiras (c|b E b|a)
        premissas_validas = divide(c, b) and divide(c, a)
        
        # Se as premissas forem válidas, verifica a conclusão (c|a)
        if premissas_validas:
            conclusao = divide(c, b+a)
            print(f"{c}|{b} e {c}|{a} ⇒ {c}|{b+a}? {conclusao} (Lineariedade válida)")
        else:
            # Mostra qual premissa falhou
            if not divide(c, b):
                print(f"{c}∤{b} (Premissa b|a inválida) ⇒ Lineariedade não aplicável")
            elif not divide(b, a):
                print(f"{c}∤{a} (Premissa b|a' inválida) ⇒ Lineariedade não aplicável")
lineariedade()

2|4 e 2|10 ⇒ 2|14? True (Lineariedade válida)
3|9 e 3|30 ⇒ 3|39? True (Lineariedade válida)
5|15 e 5|85 ⇒ 5|100? True (Lineariedade válida)
9∤49 (Premissa b|a' inválida) ⇒ Lineariedade não aplicável


##### Propriedade (iii) - Divisibilidade do Zero:  
##### $iii)\quad b \mid 0 \ \text{$\forall$ } b \in \mathbb{Z}\$

In [9]:
def divisibilidadeDoZero():
    for b in range(-2,3):
        if divide(b,0):
            conclusao = print(f"{b}|{0})? {divide(b,0)} (Divisibilidade do zero válida)")
        else:
            print("Não aplicável")
            
divisibilidadeDoZero()
        
        

-2|0)? True (Divisibilidade do zero válida)
-1|0)? True (Divisibilidade do zero válida)
0|0)? True (Divisibilidade do zero válida)
1|0)? True (Divisibilidade do zero válida)
2|0)? True (Divisibilidade do zero válida)


##### Divisibilidade por Um:  
  $iv)\quad 1 \mid a \ \text{$\forall$ } a \in \mathbb{Z}$


In [10]:
def divisibilidadePorUm():
    for a in range(-2,3):
        if divide(1,a):
            conclusao = print(f"{1}|{a} ? {divide(1,a)} (Divisibilidade por um válida)")
        else:
            print("Não aplicável")

divisibilidadePorUm()
            

1|-2 ? True (Divisibilidade por um válida)
1|-1 ? True (Divisibilidade por um válida)
1|0 ? True (Divisibilidade por um válida)
1|1 ? True (Divisibilidade por um válida)
1|2 ? True (Divisibilidade por um válida)


##### Propriedade (v) - Divisibilidade por Zero:  
  $v)\quad 0 \mid a \iff a = 0$

In [11]:
def divisibilidadePorZero():
    for a in range(-2,3):
        premissa_valida = a == 0

        if premissa_valida:
            conclusao = print(f"{0}|{a} ? {divide(a,a)} (Divisibilidade por Zero)")
        else:
            print(f"{0}|{a} ? Não aplicável, premissa a = 0 não satisfeita")

divisibilidadePorZero()
            

0|-2 ? Não aplicável, premissa a = 0 não satisfeita
0|-1 ? Não aplicável, premissa a = 0 não satisfeita
0|0 ? True (Divisibilidade por Zero)
0|1 ? Não aplicável, premissa a = 0 não satisfeita
0|2 ? Não aplicável, premissa a = 0 não satisfeita


O seguinte trecho computa todos os divisores de $121 = 11 \cdot 11$:

In [12]:
>>> n=121
>>> D121=[d for d in range(1,n+1) if n%d == 0]
>>> D121

[1, 11, 121]

Observe que o resultado é uma lista, a estrutura de dados mais importante do Python, denotada com `[` e `]`

**<span style="color: red;"> [1.3] Calcule `D48`, a lista de divisores de $48$, e para algum outro número $m < 200$. </span>**

In [13]:
>>> D48 = [d for d in range (1,49) if 48%d == 0]
>>> D48

[1, 2, 3, 4, 6, 8, 12, 16, 24, 48]

In [14]:
>>> m = 149
>>> D149 = [ d for d in range (1,m+1) if m%d == 0]
>>> D149

[1, 149]

As funções `len(L)` e `sum(L)` computam o número de elementos e a soma dos elementos de uma lista, respectivamente, então podemos computar o número de divisores, $\sum_{d \mid n} 1$, e a soma de todos os divisores, $\sum_{d \mid n} d$.

In [15]:
>>> len(D121)
3
>>> sum(D121)
133

133

 **<span style="color: red;"> [1.4] Calcule esses valores para $48$ e para $m$. </span>**

In [16]:
>>> len(D48)

10

In [17]:
>>> sum(D48)

124

In [18]:
>>> len(D149)

2

In [19]:
>>> sum(D149)

150

**<span style="color: red;"> *[1.5] Um número é perfeito se ele é igual à soma de seus divisores, excluindo ele mesmo. Então se $\sum_{d \mid n} d = 2n$. Escreva um programa para encontrar todos os números perfeitos menores que 1000.* </span>**

In [20]:
def numeroPerfeito():
    for i in range(1,1000):
        Di = [ d for d in range(1,i) if i%d==0]
        if sum(Di) == i: print("O número", i, "eh perfeito")
numeroPerfeito()

O número 6 eh perfeito
O número 28 eh perfeito
O número 496 eh perfeito


## Máximo Divisor Comum

**Definição** O máximo divisor comum de $a$ e $b$ é o maior número $d$ que é divisor de ambos $a$ e $b$.

*Notação:* $d=\mathsf{gcd}(a,b)$

**Definição** $a$ e $b$ são coprimos (ou primos entre si) se $\mathsf{gcd}(a,b)=1$.

**Teorema** A probabilidade de que $a$ e $b$ arbitrários sejam coprimos pode ser definida como
$$
\lim_{N\rightarrow\infty}\frac{|\{1\leq a,b \leq N \wedge \mathsf{gcd}(a,b)=1 \}|}{N^2} = \frac{6}{\pi^2}
$$

## Algoritmo de Euclides

**Teorema** Sejam $a,b\in\mathbb{Z}$ e $d=gcd(a,b)$. Então, $d \mid r$ tal que
$$a=q\cdot b+r\ \mathsf{com}\ 0\le r<b$$

. . .

O Algoritmo de Euclides encontrará tal $d$.

. . .

**Algoritmo** Para $a,b\in\mathbb{Z}$ tais que $a>b$,

- defina $r_0=a$,
- defina $r_1=b$,
- enquanto $r_k\neq0$
  - defina $r_k$ recursivamente como o resto de $r_{k-2}/r_{k-1}$, ou seja, $$r_{k-2}=q_{k-1} \cdot r_{k-1} + r_k\ \mathsf{com}\ 0\le r_k<r_{k-1}$$
- retorne $gcd(a,b)=r_{k-1}$

## Algoritmo de Euclides (exemplo)

**Exemplo** $a=1057$, $b=315$, $gcd(a,b)=\ ???$

$$\begin{array}{l|ccccc||ll}
& & & & & & & r_0=1057 \\
&r_{k-2} &=& q_{k-1} \cdot r_{k-1} &+& r_k & & r_1=315 \\ \hline
k=2&1057 &=& 3\cdot 315 &+& 112 & q_1=3 & r_2=112 \\
k=3& 315 &=& 2\cdot 112 &+& 91 & q_2=2 & r_3=91 \\
k=4& 112 &=& 1\cdot 91 &+& 21 & q_3=1 & r_4=21 \\
k=5& 91 &=& 4\cdot 21 &+& 7 & q_4=4 & r_5=7 \longleftarrow \\
k=6& 21 &=& 3\cdot 7 &+& 0 & q_5=3 & r_6=0
\end{array}$$

Quando $r_k  = 0$, o algoritmo termina e a resposta é $mdc(a,b) = r_{k-1} = r_5 = 7$.

O programa a seguir calcula o MDC:



In [21]:
def gcd_euclid_iterative(a, b):
# Calcula o MDC usando o algoritmo de Euclides (versão iterativa)
    while b != 0:
        (a, b) = (b, a%b) # usa uma tupla dupla
        return a

**<span style="color: red;"> [1.6] Modifique o programa para mostrar todas as etapas intermediárias. Use este programa modificado para verificar o exemplo, e para calcular o MDC entre sua matrícula e seu número de telefone. </span>**

In [22]:
def gcd_euclid_iterative(a, b):
    # Calcula o MDC usando o algoritmo de Euclides (versão iterativa), mostrando as etapas intermediárias
    r0, r1 = a, b
    
    print(f"r0 = {r0}")
    print(f"r1 = {r1}")
    print("-----------------------------------------------")
    
    i = 2

    while b != 0:
        q = a // b
        r = a % b
        
        print(f"{a} = {q}·{b} + {r}")
        print(f"r{i} = {r} | q{i-1} = {q}")
        print("-----------------------------------------------")
        (a, b) = (b, r)
        i += 1
        

    print(f"GCD({r0}, {r1}) = r{i-2} = {a}")
    return a


Verificando o exemplo: $MDC(1057,315)$

In [23]:
gcd_euclid_iterative(1057, 315)

r0 = 1057
r1 = 315
-----------------------------------------------
1057 = 3·315 + 112
r2 = 112 | q1 = 3
-----------------------------------------------
315 = 2·112 + 91
r3 = 91 | q2 = 2
-----------------------------------------------
112 = 1·91 + 21
r4 = 21 | q3 = 1
-----------------------------------------------
91 = 4·21 + 7
r5 = 7 | q4 = 4
-----------------------------------------------
21 = 3·7 + 0
r6 = 0 | q5 = 3
-----------------------------------------------
GCD(1057, 315) = r5 = 7


7

Calculando o MDC entre minha matrícula e meu número de telefone:

In [24]:
gcd_euclid_iterative(2024106034, 31971466441)

r0 = 2024106034
r1 = 31971466441
-----------------------------------------------
2024106034 = 0·31971466441 + 2024106034
r2 = 2024106034 | q1 = 0
-----------------------------------------------
31971466441 = 15·2024106034 + 1609875931
r3 = 1609875931 | q2 = 15
-----------------------------------------------
2024106034 = 1·1609875931 + 414230103
r4 = 414230103 | q3 = 1
-----------------------------------------------
1609875931 = 3·414230103 + 367185622
r5 = 367185622 | q4 = 3
-----------------------------------------------
414230103 = 1·367185622 + 47044481
r6 = 47044481 | q5 = 1
-----------------------------------------------
367185622 = 7·47044481 + 37874255
r7 = 37874255 | q6 = 7
-----------------------------------------------
47044481 = 1·37874255 + 9170226
r8 = 9170226 | q7 = 1
-----------------------------------------------
37874255 = 4·9170226 + 1193351
r9 = 1193351 | q8 = 4
-----------------------------------------------
9170226 = 7·1193351 + 816769
r10 = 816769 | q9 = 7
-------

1

**<span style="color: red;"> [1.7] Execute o programa modificado em $a=4181;b=2584$, dois números consecutivos da sequência de Fibonacci. O que você vê? Por que este é o pior caso para o algoritmo de Euclides? </span>**

In [25]:
gcd_euclid_iterative(4181, 2584)

r0 = 4181
r1 = 2584
-----------------------------------------------
4181 = 1·2584 + 1597
r2 = 1597 | q1 = 1
-----------------------------------------------
2584 = 1·1597 + 987
r3 = 987 | q2 = 1
-----------------------------------------------
1597 = 1·987 + 610
r4 = 610 | q3 = 1
-----------------------------------------------
987 = 1·610 + 377
r5 = 377 | q4 = 1
-----------------------------------------------
610 = 1·377 + 233
r6 = 233 | q5 = 1
-----------------------------------------------
377 = 1·233 + 144
r7 = 144 | q6 = 1
-----------------------------------------------
233 = 1·144 + 89
r8 = 89 | q7 = 1
-----------------------------------------------
144 = 1·89 + 55
r9 = 55 | q8 = 1
-----------------------------------------------
89 = 1·55 + 34
r10 = 34 | q9 = 1
-----------------------------------------------
55 = 1·34 + 21
r11 = 21 | q10 = 1
-----------------------------------------------
34 = 1·21 + 13
r12 = 13 | q11 = 1
-----------------------------------------------
21 = 1·13 + 8

1

**Análise do pior caso do algoritmo de Euclides com números de Fibonacci**

Ao executar o algoritmo de Euclides com os números `a = 4181` e `b = 2584`, observei que o algoritmo realiza **19 etapas** até encontrar o MDC. Esses dois números são consecutivos na sequência de Fibonacci: $ F_{19} = 4181  \text{ e}  F_{18} = 2584 $.

A cada etapa, os valores diminuem seguindo a própria sequência de Fibonacci em ordem decrescente: \( (4181, 2584) \), \( (2584, 1597) \), \( (1597, 987) \), ..., até chegar em \( (1, 0) \). Isso ocorre porque o algoritmo faz a operação `a % b`, e quando os números são consecutivos de Fibonacci, esse resto é sempre o número Fibonacci anterior.

Este é considerado o **pior caso para o algoritmo de Euclides** porque o número de passos é **máximo** em relação ao tamanho dos números de entrada. O número de etapas cresce linearmente com a posição de Fibonacci, enquanto em casos comuns o número de passos cresce proporcional ao logaritmo dos números. Ou seja, para dois números tão grandes, o algoritmo normalmente terminaria mais rápido — mas com números de Fibonacci, ele faz o **número máximo possível de chamadas**.


**<span style="color: red;"> =============================================================================================</span>**

# Números primos

**Definição** $n$ é primo se tiver apenas dois divisores, a saber, $1$ e ele mesmo.

**<span style="color: red;"> [2.1] Escreva um teste em Python para verificar se $n$ é primo. </span>**

In [26]:
import math
def ehPrimo(n):
    for d in range(2,round(math.sqrt(n)+1)):
        if (n % d==0):
            print("Não eh primo, divisor encontrado:", d)
            return False
    print("Eh primo")
    return True

ehPrimo(65537)

Eh primo


True

**Teorema** Existem infinitos primos.

**Demonstração (Euclides)** Considere $n=(2\cdot 3\cdot 5\cdot \ldots\cdot p_{max})+1$. Certamente $n>p_{max}$ e $p_i\not\mid\n$ para todo $i$ (já que $p_i \mid n-1$). Portanto, ou $n$ é primo, ou o produto de primos maiores que $p_{max}$.

### Fatos sobre números primos

**Definição** Um primo de Mersenne é um primo da forma $M_k=2^k-1$

$$\begin{array}{c|ccccccccc}
k & 2 & 3 & 5 & 7 & 13 & 17 & 19 & 31 & \ldots \\ \hline
2^k & 4 & 8 & 32 & 128 & 8192 & 131072 & 524288 & 2147483648 & \ldots \\
M_k & 3 & 7 & 31 & 127 & 8191 & 131071 & 524287 & 2147483647 & \ldots
\end{array}$$

Existe uma competição mundial para encontrar o maior primo; sempre da forma $M_k=2^k-1$. São mais simples de encontrar porque: $M_k\ \mathsf{primo} \stackrel{\not\Leftarrow}{\Rightarrow}k\ \mathsf{primo}$. Caso contrário, os primos de Mersenne não têm importância prática.

**<span style="color: red;"> [2.2] Encontre na internet qual é o maior número primo conhecido no momento. É um primo de Mersenne? </span>**

O maior número primo conhecido é o número $2^{136279841} - 1$.  
Percebe-se que este número segue a forma $2^{k} - 1$, portanto, é um primo de $Mersene$: $M_{136279841} = 2^{136279841} - 1$.


**Definição** Um primo de Fermat é um primo da forma $F_k = 2^{2^k}+1$

$$\begin{array}{c|cccccc}
k & 1 & 2 & 3 & 4 & 5 & \ldots \\ \hline
2^k & 2 & 4 & 8 & 16 & 32 & \ldots \\
F_k & 5 & 17 & 257 & 65537 & 4294967297 & \ldots
\end{array}$$

- Fermat observou que $F_1,F_2,F_3,F_4$ são primos e conjecturou que todos os outros eram primos.

- De fato, nenhum primo de Fermat $F_k$ foi encontrado para $k>4$.

- Os primos de Fermat são usados em criptografia devido à sua representação binária simples.

**<span style="color: red;"> [2.3] Qual é a representação binária de $257$ e de $65537$? Você consegue encontrar um comando em Python que faça isso para você </span>**

Função para encontrar a representação binária de qualquer número Inteiro:

In [27]:
def representacaoBinaria(n):
    num=n
    numeros = []
    #print(n)
    while n!=1:
        resto = n%2
        n = n//2
        numeros.insert(0, resto)
        #print(resto, n) # Imprime as etapas intermediárias
    numeros.insert(0, n)
    print(f"O número {num} em binário é:", end="")
    print("".join(str(item) for item in numeros))

Encontrando a representação binária de 257 e 65537 usando o algoritmo acima:

In [28]:
representacaoBinaria(257)

O número 257 em binário é:100000001


In [29]:
representacaoBinaria(65537)

O número 65537 em binário é:10000000000000001


Encontrando a representação binária de 257 e 65537 usando a função bin do python:

In [30]:
print(bin(257)[2:])

100000001


In [31]:
print(bin(65537)[2:])

10000000000000001


## Quantos números primos existem?

**Definição** A função $\pi(x)$ conta todos os números primos até $x$.

$$\pi(x) = | \{ 1<p\leq x : p \ \mathsf{é\ primo}\}|$$

**Exemplos** $\pi(10)=|\{2,3,5,7\}|= 4$

A expressão $\pi(x)/x$ representa a densidade de primos entre todos os inteiros.

**Teorema dos Números Primos**
$$\lim_{N\rightarrow\infty}\frac{\pi(N)}{N}\sim\frac{1}{\ln(N)}$$

A demonstração deste teorema é muito difícil.

![](PrimeNumberTheorem.png)

**<span style="color: red;"> [2.4] Use este teorema para estimar quantos números primos existem com tamanhos de 1024 bits, 1536 bits e 2048 bits (os tamanhos usados na criptografia) </span>**

In [32]:
import math

def eh_primo(n):
    for d in range(2,round(math.sqrt(n)+1)):
        if (n % d==0):
            return False
    return True

def pi(x):
    count = 0
    for i in range(2, x+1):
        if eh_primo(i):
            count +=1
    return count

def verificarTeorema(x):
    pi_x = pi(x)
    aprox = x / math.log(x) if x > 1 else 0
    print(f"π({x}) = {pi_x}")
    print(f"Aproximação pelo Teorema dos Números Primos: {aprox:.2f}")
    print(f"Razão π({x})/({x}) = {pi_x/x:.5f}")
    print(f"1/ln({x}) = {1/math.log(x):.5f}\n")

from decimal import Decimal, getcontext

getcontext().prec = 1000

def estimarPrimos(x):
    x_decimal = Decimal(x)
    aprox = x_decimal / x_decimal.ln() if x_decimal > 1 else 0
    print(f"Aproximação pelo Teorema dos Números Primos para x = {x_decimal:.2e}:\n≈ {aprox:.2e}\n")

valores = [2**1024, 2**1536, 2**2048]


for v in valores:
    estimarPrimos(v)


Aproximação pelo Teorema dos Números Primos para x = 1.80e+308:
≈ 2.53e+305

Aproximação pelo Teorema dos Números Primos para x = 2.41e+462:
≈ 2.26e+459

Aproximação pelo Teorema dos Números Primos para x = 3.23e+616:
≈ 2.28e+613



**Teorema de Fermat** Se $p$ é primo, então $a^{p-1}\equiv 1 \pmod{p}$ para todo $a \in [1,p-1]$.

Exemplo: $2^{10} = 1024 = 93 \cdot 11 + 1\equiv 1 \pmod{11}$


In [33]:
>>> pow(2,10)

1024

In [34]:
>>> pow(2,10,11)

1

**<span style="color: red;"> [2.5] Use Python para testar o Teorema de Fermat para outros valores de $a$ e $p$. </span>**

$3^{6} = 729 = 104 \cdot 7 + 1\equiv 1 \pmod{7}$

In [35]:
>>> pow(3, 6)

729

In [36]:
>>> pow(3, 6, 7)

1

$4^{4} = 256 = 51 \cdot 5 + 1\equiv 1 \pmod{5}$

In [37]:
>>> pow(4, 4)

256

In [38]:
>>> pow(4, 4, 5)

1

O inverso do teorema de Fermat pode ser usado como um teste de primalidade, pois se $n$ não satisfaz a equação para algum $a$, então não pode ser primo.

**Contra-positivo** Se $a^{n-1}\not\equiv 1 \pmod{n}$ para algum $a \in [1,n-1]$, então $n$ é composto.



**Teste de primalidade de Fermat** Para $n,k\in\mathbb{N}$,

- repita $k=10$ vezes:
- escolha aleatoriamente $a \in [2,n-2]$
- defina $b\equiv a^{n-1} \pmod{n}$
- se $b \neq 1$ retorne **composto** (com testemunha $a$)
- retorne **(provavelmente) primo**

O teste de primalidade de Fermat não fornece certeza absoluta sobre $n$ ser primo ou não, pois alguns números são imunes a esse teste.

**Definição** Um número de Carmichael é um composto $n$ tal que ${a^{n-1}\equiv 1\pmod{n}}$ para todo $a \in [2,n-1]$.

**Exemplo** $561 = 3\cdot 11\cdot 17$, ou $41041 = 7\cdot 11\cdot 13$

Mas os números de Carmichael são raros. Até 10000 são apenas 7, a saber $561, 1105, 1729, 2465, 2821, 6601, 8911$. A probabilidade de por acaso testar um número de Carmichael é muito pequena, então não nos preocuparemos com eles.

Isso nos leva ao seguinte programa para primalidade, que você pode usar na próxima questão.

In [39]:
import random

def fermat_primality_test(n, k=10):
		"""
		Teste de primalidade de Fermat para verificar se n é provavelmente primo.

		Argumentos:
				n (int): O número a ser testado.
				k (int): O número de testes aleatórios a serem realizados (padrão=10).

		Retorna:
				bool: Verdadeiro se n passar em todos os testes (provavelmente primo), Falso se composto.
"""
		if n <= 1:
				return False
		elif n <= 3:
				return True # 2 e 3 são primos
		elif n % 2 == 0:
				return False # Números pares >2 são compostos

		for _ in range(k):
				a = random.randint(2, n - 2) # Escolha um valor aleatório a onde 1 < a < n-1
				if pow(a, n - 1, n) != 1: # Calcula a^(n-1) mod n eficientemente
						return False # Definitivamente composto
				return True # Provavelmente primo

**<span style="color: red;"> [2.6] Construa um número primo com seu nome </span>**

1. Crie uma frase com cerca de 45 letras, incluindo seu nome. Cada aluno de uma dupla/tripla deve fazer isso.
2. Converta para um número como segue: $A=01, ..., Z=26, \mathsf{space}=27$ etc. Concatenando os dígitos, você obterá um número com cerca de 90 dígitos.
3. Adicione 10 dígitos aleatórios no final para obter um número ímpar de 100 dígitos $n$ e teste se $n$ é um número primo. Caso contrário, altere os últimos 10 dígitos até que $n$ seja primo.
4. Você pode usar `fermat_primality_test(n, k=10)`.

In [40]:
import random

# Funções auxiliares (letra_para_numero, frase_para_numero)
def letra_para_numero(c):
    if c == ' ':
        return '27'
    else:
        return f"{ord(c.upper()) - ord('A') + 1:02d}"

def frase_para_numero(frase):
    return ''.join(letra_para_numero(c) for c in frase)

frase = "MARNEY MELO GOSTA DE ESTUDAR OS NUMEROS PRIMOS"
base_numero = frase_para_numero(frase)

# Adicionar 10 dígitos aleatórios e testar até encontrar um primo.
while True:
    sufixo_aleatorio = ''.join(random.choice('0123456789') for _ in range(9))
    ultimo_digito_impar = random.choice('1379')
    
    numero_candidato = int(base_numero + sufixo_aleatorio + ultimo_digito_impar)
    
    if fermat_primality_test(numero_candidato, k=10):
        print(f"Número gerado: {numero_candidato}")
        print(f"Número de dígitos: {len(str(numero_candidato))}")
        print("Provavelmente primo (segundo o Teste de Fermat)")
        break

Número gerado: 130118140525271305121527071519200127040527051920210401182715192714211305181519271618091315192279029529
Número de dígitos: 102
Provavelmente primo (segundo o Teste de Fermat)


**<span style="color: red;"> =============================================================================================</span>**

# Classes de congruência



**Definição** Seja $m$ um inteiro. Dizemos que $a$ é congruente a $b$ módulo $m$, ou que $a$ e $b$ pertencem à mesma classe de congruência, se $a-b$ for múltiplo de $m$, ou seja, existe $k \in \mathbb{Z}$ tal que
$$
a-b = k\cdot m
$$
Notação matemática:
$$
a \equiv b \pmod m
$$

Em outras palavras, significa que o resto módulo $m$ é o mesmo em ambos os casos. Em Python, C e Python, `%` é o operador mod, então isso significa que o teste
```
a % m == b % m
```
retorna `True`.

Usamos aritmética modular no dia a dia:

- $8272107104 \equiv 8274104 \pmod {1000}$ apenas verificando os três últimos dígitos.
- $2 \equiv 14 \pmod {12}$, então ambos representam 2 horas, de noite ou de tarde.
- $3 \equiv 24 \mod{7}$, então o 3º e o 24º dia de um mês caem no mesmo dia da semana
- Curiosidade: as datas 04/04, 06/06, 08/08, 10/10 e 12/12 de um ano caem todos no mesmo dia da semana; em 2025, esta é uma sexta-feira. Isso ocorre porque, em todos os casos, o número de dias entre duas datas é $30+31+2=63$ ou $31+30+2=63$, um múltiplo de $7$. Também temos que 09/05, 05/09 e 07/11 e 11/07 caem no mesmo dia da semana. Março e fevereiro são mais complicados por causa dos anos bissextos. Mas o último dia de fevereiro, matematicamente conhecido como 00/03, também cai neste dia. Se não for um ano bissexto, o último dia de janeiro, 31/01, cai no mesmo dia.



### Adição modular

É possível realizar a adição módulo $n$:
$$
a \text{\ mod\ } n+b \text{\ mod\ } n = (a+b) \text{\ mod\ } n
$$
Observe que obtemos as propriedades de um grupo Abeliano:

- **fechado** $\forall\ a,b \in G: a + b \in \mathbb{Z}_n^*$

- **associativo** $\forall\ a,b,c \in \mathbb{Z}_n^*: (a + b) + c = a + (b + c)$

- **elemento neutro de existência** $\forall\ a \in \mathbb{Z}_n^*: a + 0 = 0 + a = a$

- **elemento inverso de existência** $\forall\ a \in \mathbb{Z}_n^*\ \exists\ -a: a + (-a) = (-a) + a = 0$

- **comutativo** $\forall\ a,b \in \mathbb{Z}_n^*: a + b = b+a$

A adição modular é usada em criptografia.

### Exemplo: Cifra de Vigenère usando o grupo aditivo módulo $n$

**Contexto:**

1. **Grupo aditivo $\mathbb{Z}_n$**
- O conjunto $\{0, 1, 2, \dots, n-1\}$ sob adição módulo $n$.
- Exemplo: Em $\mathbb{Z}_{26}$, $25 + 3 = 2 \mod 26$.

2. **Cifra de Vigenère**
- **Criptografia**: $C_i = (P_i + K_i) \mod 26$
- **Descriptografia**: $P_i = (C_i - K_i) \mod 26$
- Aqui, $P_i, K_i, C_i$ são representações numéricas de letras de texto simples, chave e texto cifrado (A=0, B=1, ..., Z=25).

**Parte 1: Cifrar**

1. **Texto simples**: `"CRYPTO"`
- Converta para números: $C=2, R=17, Y=24, P=15, T=19, O=14$.
2. **Chave**: `"KEY"` (repetido como `"KEYKEY"`).
- Converta para números: $K=10, E=4, Y=24$.
3. **Criptografe** usando $C_i = (P_i + K_i) \mod 26$:
- $(2+10) \mod 26 = 12$ → M
- $(17+4) \mod 26 = 21$ → V
- $(24+24) \mod 26 = 48 \mod 26 = 22$ → W
- $(15+10) \mod 26 = 25$ → Z
- $(19+4) \mod 26 = 23$ → X
- $(14+24) \mod 26 = 38 \mod 26 = 12$ → M
4. **Texto cifrado**: `"MVWZXM"`.

**<span style="color: red;">[3.1] Verifique este exemplo usando Python. Use as seguintes funções para converter entre letras e números</span>**

In [41]:
# Versão  usando string e index()
def numero_para_letra_v2(numero):
    """Converte um número (0-25) para letra (A-Z) usando string e index."""
    letras = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
    if not 0 <= numero <= 25:
        raise ValueError("O número deve estar entre 0 e 25.")
    return letras[numero]

def letra_para_numero_v2(letra):
    """Converte uma letra (A-Z) para número (0-25) usando string e index()."""
    letras = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
    if not letra.isalpha() or not letra.isupper() or len(letra) != 1:
        raise ValueError("A entrada deve ser uma única letra maiúscula (A-Z).")
    return letras.index(letra)

def cifragem_vigenere(mensagem_enviada, chave_cifra):
    index = 0;
    mensagem_criptografada = []
    for c in mensagem_enviada:
        numero_cripto = letra_para_numero_v2(c) + letra_para_numero_v2(chave_cifra[index % len(chave_cifra)])
        mensagem_criptografada.append(numero_para_letra_v2(numero_cripto % 26))
        index+=1
    return "".join(mensagem_criptografada)

**Testando algumas possibilidades de criptografia:**

**1.1 Exemplo que foi dado pelo roteiro:**

In [42]:
print("Mensagem criptografada:", cifragem_vigenere("CRYPTO", "KEY"))

Mensagem criptografada: MVWZXM


**1.2 Vamos testar com qualquer tamanho de palavra e chave:**

In [43]:
print("Mensagem criptografada:", cifragem_vigenere("MENSAGEMSUPERSECRETA", "CHAVECIFRA"))

Mensagem criptografada: OLNNEIMRJURLRNIEZJKA


**1.2.1 E se tivermos uma palavra menor que a chave, a criptografia ainda funciona?**

In [44]:
print("Mensagem criptografada:", cifragem_vigenere("MPEQUENA", "CHAVECIFRAGIGANTE"))

Mensagem criptografada: OWELYGVF


**Parte 2: Decifrar**
1. **Texto cifrado**: `"MVWZXM"` → $M=12, V=21, W=22, Z=25, X=23, M=12$.
2. **Mesma chave**: `"KEYKEY"` → $K=10, E=4, Y=24$.
3. **Descriptografe** usando $P_i = (C_i - K_i) \mod 26$:
- $(12-10) \mod 26 = 2$ → C
- $(21-4) \mod 26 = 17$ → R
- $(22-24) \mod 26 = -2 \mod 26 = 24$ → Y
- $(25-10) \mod 26 = 15$ → P
- $(23-4) \mod 26 = 19$ → T
- $(12-24) \mod 26 = -12 \mod 26 = 14$ → O
4. **Texto Simples Recuperado**: `"CRYPTO"`.

**<span style="color: red;">[3.2] Verifique este exemplo de descriptografia.</span>**

In [45]:
def descriptografia_vigenere(mensagem_criptografada, chave_cifra):
    index = 0;
    mensagem_descriptografada = []
    for c in mensagem_criptografada:
        numero_descripto = letra_para_numero_v2(c) - letra_para_numero_v2(chave_cifra[index % len(chave_cifra)])
        mensagem_descriptografada.append(numero_para_letra_v2(numero_descripto % 26))
        index+=1
    return "".join(mensagem_descriptografada)
        

**Vamos testar a descriptografia do exemplo assim como pede o relatório, mas também, vamos analisar se a criptografia realizada anteriormente, em todos os exemplos, é verdadeira:**

In [46]:
print ("Mensagem original:", descriptografia_vigenere("MVWZXM", "KEY"))

Mensagem original: CRYPTO


In [47]:
print ("Mensagem original:", descriptografia_vigenere("OWELYGVF", "CHAVECIFRAGIGANTE"))

Mensagem original: MPEQUENA


**No caso a seguir, fica claro que a mensagem original é a mesma que foi passada como parâmetro para ser criptografada e, logo em sequência, ser descriptografada. Dessa forma, podemos ver que nossa função funciona corretamente.**

In [48]:
print ("Mensagem original:", descriptografia_vigenere(cifragem_vigenere("MENSAGEMSUPERSECRETA", "CHAVECIFRA"), "CHAVECIFRA"))

Mensagem original: MENSAGEMSUPERSECRETA


**<span style="color: red;">[3.3]Desafio Bônus: Tente quebrar este texto cifrado (use $\mathbb{Z}_{26}$)</span>**:

- Texto cifrado: `"MOBIAKCKVGCNNJSJZRIB"`
- Comprimento da chave conhecido: 2 (ex.: `"AB"`).
- Texto simples: `"MNBHAJCJVFCMNRJYRHB"` (provavelmente chave errada; demonstra tentativa e erro).
- Tente outras chaves até encontrar uma palavra em inglês.

In [49]:
print ("Mensagem original:", descriptografia_vigenere("MOBIAKCKVGCNNJSJZRIB", "CD"), " - NÃO SERVE")
print ("Mensagem original:", descriptografia_vigenere("MOBIAKCKVGCNNJSJZRIB", "AB"), " - NÃO SERVE")
print ("Mensagem original:", descriptografia_vigenere("MOBIAKCKVGCNNJSJZRIB", "RL"), " - NÃO SERVE")
print ("Mensagem original:", descriptografia_vigenere("MOBIAKCKVGCNNJSJZRIB", "MD"), " - NÃO SERVE")

Mensagem original: KLZFYHAHTDAKLGQGXOGY  - NÃO SERVE
Mensagem original: MNBHAJCJVFCMNISIZQIA  - NÃO SERVE
Mensagem original: VDKXJZLZEVLCWYBYIGRQ  - NÃO SERVE
Mensagem original: ALPFOHQHJDQKBGGGNOWY  - NÃO SERVE


**Por quê não automatizar a busca?**

In [50]:
for a in range(0,26):
    for b in range(0,26):
        aux = str()
        aux += numero_para_letra_v2(a)
        aux += numero_para_letra_v2(b)
        print(descriptografia_vigenere("MOBIAKCKVGCNNJSJZRIB",aux), "Senha usada:", aux)

MOBIAKCKVGCNNJSJZRIB Senha usada: AA
MNBHAJCJVFCMNISIZQIA Senha usada: AB
MMBGAICIVECLNHSHZPIZ Senha usada: AC
MLBFAHCHVDCKNGSGZOIY Senha usada: AD
MKBEAGCGVCCJNFSFZNIX Senha usada: AE
MJBDAFCFVBCINESEZMIW Senha usada: AF
MIBCAECEVACHNDSDZLIV Senha usada: AG
MHBBADCDVZCGNCSCZKIU Senha usada: AH
MGBAACCCVYCFNBSBZJIT Senha usada: AI
MFBZABCBVXCENASAZIIS Senha usada: AJ
MEBYAACAVWCDNZSZZHIR Senha usada: AK
MDBXAZCZVVCCNYSYZGIQ Senha usada: AL
MCBWAYCYVUCBNXSXZFIP Senha usada: AM
MBBVAXCXVTCANWSWZEIO Senha usada: AN
MABUAWCWVSCZNVSVZDIN Senha usada: AO
MZBTAVCVVRCYNUSUZCIM Senha usada: AP
MYBSAUCUVQCXNTSTZBIL Senha usada: AQ
MXBRATCTVPCWNSSSZAIK Senha usada: AR
MWBQASCSVOCVNRSRZZIJ Senha usada: AS
MVBPARCRVNCUNQSQZYII Senha usada: AT
MUBOAQCQVMCTNPSPZXIH Senha usada: AU
MTBNAPCPVLCSNOSOZWIG Senha usada: AV
MSBMAOCOVKCRNNSNZVIF Senha usada: AW
MRBLANCNVJCQNMSMZUIE Senha usada: AX
MQBKAMCMVICPNLSLZTID Senha usada: AY
MPBJALCLVHCONKSKZSIC Senha usada: AZ
LOAIZKBKUGBNMJRJYRHB Senha usada: BA
L

**Podemos perceber que a senha "JG" resulta em uma palavra em inglês:**

In [51]:
print(descriptografia_vigenere("MOBIAKCKVGCNNJSJZRIB","JG"), "Senha usada:", "JG")

DISCRETEMATHEDJDQLZV Senha usada: JG


**Principais Conclusões**
- Vigenère usa **aritmética modular** ($\mathbb{Z}_{26}$) para generalizar deslocamentos.
- A segurança depende do **comprimento da chave** e da **aleatoriedade**.
- O grupo aditivo $\mathbb{Z}_n$ é fundamental para cifras clássicas.

### Como fica no mundo digital, dos bits 0 e 1

**<span style="color: red;">[3.4] Produza uma tabela para todas as quatro adições possíveis módulo 2. Compare-a com uma tabela da operação ou-exclusivo (xor). O que você vê?</span>**


$$
\begin{array}{c|cc}
+&0&1\\ \hline
0&0&1 \\
1&1&0\\
\end{array}
$$


$$
\begin{array}{c|cc}
\oplus&0&1\\ \hline
0&0&1 \\
1&1&0 \\
\end{array}
$$

**Pode-se notar que as tabelas são equivalentes**

# Produto direto de grupos aditivos

**Definição** Sejam $G_1$ e $G_2$ dois grupos. Então, o produto direto $G$ tem

- defina $$G_1 \times G_2 = \{(g_1,g_2)|g_1 \in G_1 \wedge g_2 \in G_2\}$$;

- operação $(g_1,g_2) \cdot (h_1,h_2) := (g_1 h_1, g_2 h_2)$

**Exemplo**

$\mathbb{Z}_3 \times \mathbb{Z}_5 = \{(g,h) \mid 0 \leq g < 3 \wedge 0 \leq h < 5 \}$

$(2,4)+(0,3)=(2+0\!\pmod 3, 4+3 \!\pmod 5)=(2,2)$

Na verdade, $\mathbb{Z}_3 \times \mathbb{Z}_5$ é equivalente (exceto pela notação) a $\mathbb{Z}_{15}$:
$$
\mathbb{Z}_3 \times \mathbb{Z}_5 = \mathbb{Z}_{15}
$$

$$
\begin{array}{c|ccccc}
&0&1&2&3&4\\ \hline
0&0=(0,0)&6=(0,1)&12=(0,2)&3=(0,3)&9=(0,4) \\
1&10=(1,0)&1=(1,1)&7=(1,2)&13=(1,3)&4=(1,4) \\
2&5=(2,0)&11=(2,1)&2=(2,2)&8=(2,3)&14=(2,4) \\
\end{array}
$$


**<span style="color: red;">[3.5] Produza uma tabela semelhante para n = 35</span>**


**Precisamos de uma tabela para n = 35, e para fazer isso, basta multiplicar $\mathbb{Z}_5 \times \mathbb{Z}_7$, posto que ambos são coprimos e, se multiplicados, são igual a 35.**

$$
\begin{array}{c|ccccc}
&0&1&2&3&4&5&6\\ \hline
0&0=(0,0)&15=(0,1)&30=(0,2)&10=(0,3)&25=(0,4)&5=(0,5)&20=(0,6) \\
1&21=(1,0)&1=(1,1)&16=(1,2)&31=(1,3)&11=(1,4)&26=(1,5)&6=(1,6) \\
2&7=(2,0)&22=(2,1)&2=(2,2)&17=(2,3)&32=(2,4)&12=(2,5)&27=(2,6) \\
3&28=(3,0)&8=(3,1)&23=(3,2)&3=(3,3)&18=(3,4)&33=(3,5)&13=(3,6) \\
4&14=(4,0)&29=(4,1)&9=(4,2)&24=(4,3)&4=(4,4)&19=(4,5)&34=(4,6) \\
\end{array}
$$

# Outro exemplo: 11 x 13 = 143

![Captura de tela 2021-07-01 140847](.\Screenshot 2021-07-01 140847.png)

Vimos que $\mathbb{Z}_{143}^+ \cong \mathbb{Z}_{11}^+ \times \mathbb{Z}_{13}^+$. Isso pode ser generalizado:

**Teorema** Sejam $n$ e $m$ coprimos. Então
$$
\mathbb{Z}_{nm}^+ \cong \mathbb{Z}_n^+ \times \mathbb{Z}_m^+
$$
significando que ambos os grupos são isomorfos. Ou seja, os grupos são idênticos, exceto pela notação usada nos elementos.

Outra maneira de afirmar isso é por meio do teorema chinês do resto.

# Teorema Chinês do Resto

**Intuição** Suponha que se saiba que $a \equiv 3 \!\pmod {11}$ e $a \equiv 5 \!\pmod {13}$. Existe apenas um valor $\pmod {143}$ que satisfaz ambas as equações. A tabela acima mostra que esse valor é $135$.

Esta afirmação pode ser generalizada.

### **Teorema Chinês do Resto (TCR)**
O Teorema Chinês do Resto afirma que, para um sistema de módulos **coprimos em pares** $n_1, n_2, \dots, n_k$ e números inteiros dados $a_1, a_2, \dots, a_k$, existe uma **solução única** $x$ módulo $N = n_1 \cdot n_2 \cdots n_k$ para o sistema de congruências:

$$
\begin{cases}
x \equiv a_1 \mod n_1 \\
x \equiv a_2 \mod n_2 \\
\vdots \\
x \equiv a_k \mod n_k
\end{cases}
$$

Além disso, se $x$ é uma solução, então toda solução é congruente a $x \mod N$.

---

### **Problema de Exemplo**
Encontre um inteiro $x$ tal que:
$$
\begin{cases}
x \equiv 2 \mod 3 \\
x \equiv 3 \mod 5 \\
x \equiv 2 \mod 7
\end{cases}
$$

#### **Solução Passo a Passo**


1. **Verifique se os módulos são primos entre si**:
$mdc(3,5) = 1$, $mdc(3,7) = 1$, $mdc(5,7) = 1$.
→ O TCR se aplica, pois todos os módulos são primos entre si.

2. **Calcule $N = 3 \times 5 \times 7 = 105$**.

3. **Encontre soluções parciais**:
- Para $x \equiv 2 \mod 3$:
- Seja $x = 5 \times 7 \times k_1 = 35k_1$.
- Use tentativa e erro para encontrar um múltiplo de 35 que seja $2 \mod 3$. Ou seja, encontre $k_1 \in \{1,2\}$ tal que $35k_1 \equiv 2 \mod 3$.
- $k_1 = 1$ está correto, já que $35 k_1 \equiv 25 \equiv 2 \mod3$
- A solução parcial é $x_1 = 35 \times 1 = 35$.
- Observe que $35 = (2,0,0) \mod {(3,5,7)}$

- Para $x \equiv 3 \mod 5$:
- Seja $x = 3 \times 7 \times k_2 = 21k_2$.
- Use tentativa e erro para encontrar um múltiplo de 21 que seja $3 \mod 5$. Ou seja, encontre $k_2 \in \{1,2,3,4\}$ tal que $21k_2 \equiv 3 \mod 5$.
- $k_2 = 3$ é aceitável, já que $21 k_2 = 21 \cdot 3 = 63 \equiv 3 \mod 5$
- A solução parcial é $x_2 = 21 \times 3 = 63$.

- Observe que $63 = (0,3,0) \mod {(3,5,7)}$

- Para $x \equiv 2 \mod 7$:
- Seja $x = 3 \times 5 \times k_3 = 15k_3$.
- Encontre um múltiplo de 15 que seja $2 \mod 7$.
- Resolva $15k_3 \equiv 2 \mod 7$.
- Como $15 \equiv 1 \mod 7$, temos $k_3 \equiv 2 \mod 7$.
- Portanto, $k_3 = 2 + 7m$, e a solução parcial é $x_3 = 15 \times 2 = 30$.
- Observe que $30 = (0,0,2) \mod {(3,5,7)}$

4. **Combine as soluções parciais**:
A solução geral é:
$$
x \equiv x_1 + x_2 + x_3 \mod 105
$$
$$
x=(2,0,0)+(0,3,0)+(0,0,2) \mod {(3,5,7)}
$$
$$
x \equiv 35 + 63 + 30 \mod 105
$$
$$
x \equiv 128 \mod 105 \implies
$$
$$
x \equiv 23 \mod 105
$$

5. **Verificação**:
- $23 \mod 3 = 2$ ✓
- $23 \mod 5 = 3$ ✓
- $23 \mod 7 = 2$ ✓

#### **Resposta final**:
A menor solução positiva é $x = 23$. Todas as soluções têm a forma $x = 23 + 105k$ para o inteiro $k$.


**<span style="color: red;">[3.6] Use Python para encontrar uma solução para
$$
\begin{cases}
x \equiv 5 \mod 7 \\
x \equiv 4 \mod 11 \\
x \equiv 10 \mod 13
\end{cases}
$$
Verifique sua resposta.</span>**

In [52]:
import math
def tcr(a, amod, b, bmod, c, cmod):
    if math.gcd(amod, bmod) == 1 and math.gcd(amod, cmod) == 1 and math.gcd(bmod, cmod) == 1:
        mdc = amod * bmod * cmod
        xa = 0
        xb = 0
        xc = 0
        for k in range(1, amod):
            xa = bmod*cmod*k
            if xa % amod == a:
                print("Solução Parcial I:", xa)
                break;
        for k in range(1, bmod):
            xb = amod*cmod*k
            if xb % bmod == b:
                print("Solução Parcial II:", xb)
                break;
        for k in range(1, cmod):
            xc = amod*bmod*k
            if xc % cmod == c:
                print("Solução Parcial III:", xc)
                break;
        return (xa + xb + xc) % mdc


**Primeiro, vamos ver se a nossa função do Teorema Chinês do Resto funciona, usando o exemplo já dado pelo roteiro:**

In [53]:
print("Resposta final: x =", tcr(2, 3, 3, 5, 2, 7))

Solução Parcial I: 35
Solução Parcial II: 63
Solução Parcial III: 30
Resposta final: x = 23


**Agora, vamos resolver**
$$
\begin{cases}
x \equiv 5 \mod 7 \\
x \equiv 4 \mod 11 \\
x \equiv 10 \mod 13
\end{cases}
$$

In [55]:
print("Resposta final: x =", tcr(5, 7, 4, 11, 10, 13))

Solução Parcial I: 572
Solução Parcial II: 455
Solução Parcial III: 231
Resposta final: x = 257



### **Principais Conclusões**
- A TCR garante uma **solução única** módulo $N$ se os módulos forem primos entre si.
- Usado em criptografia (por exemplo, descriptografia RSA), hashing e cálculos rápidos.

**<span style="color: red;"> =============================================================================================</span>**

# Grupo Multiplicativo Módulo $N$
### **Grupos cíclicos**

Um grupo finito é chamado de **cíclico** se puder ser gerado inteiramente por um único elemento, chamado de **gerador**. Isso significa que cada elemento do grupo pode ser expresso como aplicações repetidas da operação de grupo (adição ou multiplicação) no gerador.

**Definição**
Um grupo $(G, \circ)$ é **cíclico** se existe um elemento $g \in G$ (o gerador) tal que todos os elementos podem ser obtidos aplicando a operação de grupo repetidamente em $g$, onde:

- Para **grupos aditivos** (como $\mathbb{Z}_n$), isso significa a adição de $g$ consigo mesmo $n$ vezes $\underbrace{g + g + \dots + g}_{n \text{ vezes}} = n \cdot g$; ou seja, multiplicação.

- Para **grupos multiplicativos** (como $(\mathbb{Z}_n^*$), isso significa a adição de $g$ consigo mesmo $n$ vezes $\underbrace{g \cdot g \cdot \ldots \cdot g}_{n \text{ vezes}} = g^n$; ou seja, exponenciação.

##### **Exemplo: Grupo Aditivo $\mathbb{Z}_{35}^+$**

- **Elementos**: $\{0, 1, 2, \dots, 34\}$.
- **Operação**: Adição módulo 35.
- **Cíclico?** **Sim**, pois é gerada por $1$:
$$
1 + 1 + \dots + 1 \ (\text{35 vezes}) \equiv 0 \mod 35.
$$
Todo elemento $k \in \mathbb{Z}_{35}$ pode ser escrito como $k \cdot 1 \mod 35$.
- **Geradores**: Todos os números que são coprimos com 35 (ou seja, $\text{mdc}(k, 35) = 1$):
$$
\text{Geradores} = \{1, 2, 3, 4, 6, 8, 9, 11, 12, 13, 16, 17, 18, 19, 22, 23, 24, 26, 27, 29, 31, 32, 33, 34\}.
$$

Por exemplo:
- $2$ é um gerador porque seus múltiplos abrangem todos os elementos: $2, 4, 6, \cdots, 34, 1, 3, 5, \cdots$.
- $5$ **não** é um gerador porque $\text{mdc}(5, 35) = 5 \neq 1$, e gera apenas $\{0, 5, 10, 15, 20, 25, 30\}$.

**<span style="color: red;">
[4.1] Use Python para encontrar todos os geradores de $\mathbb{Z}_{36}^+$ e de $\mathbb{Z}_{37}^+$.
    </span>**

In [56]:
import math
def gerador(n):
    geradores = []
    for g in range(0, n):
        if math.gcd(g, n) == 1:
            geradores.append(g)
    return geradores

**Vamos testar a nossa função com o exemplo dado, $\mathbb{Z}_{35}^+$. Logo em seguida, vamos testar para $\mathbb{Z}_{36}^+$ e $\mathbb{Z}_{37}^+$:**

In [57]:
print("Geradores:", gerador(35))

Geradores: [1, 2, 3, 4, 6, 8, 9, 11, 12, 13, 16, 17, 18, 19, 22, 23, 24, 26, 27, 29, 31, 32, 33, 34]


In [58]:
print("Geradores:", gerador(36))

Geradores: [1, 5, 7, 11, 13, 17, 19, 23, 25, 29, 31, 35]


In [59]:
print("Geradores:", gerador(37))

Geradores: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36]


A **função phi de Euler** (ou **função totiente de Euler**), denotada como **φ(n)**, conta o número de inteiros até **n** que são **coprimos com n** (ou seja, inteiros $1 \leq a \leq n$ com $\text{mdc}(a, n) = 1$).

**Fórmula**
Para um inteiro positivo $n$ com fatoração em primos:
$$
n = p_1^{k_1} p_2^{k_2} \cdots p_m^{k_m},
$$

A função totiente é:

$$
\varphi(n) = n \prod_{p \mid n} \left(1 - \frac{1}{p}\right).
$$

**Exemplos**
- $\varphi(10) = 4$ (coprimos: 1, 3, 7, 9).
- $\varphi(7) = 6$ (todos os números de 1 a 6, já que 7 é primo).

**<span style="color: red;">
    [4.2] Calcule os valores de $\varphi(35),\varphi(36),\varphi(37)$ e compare com os resultados obtidos acima.
    </span>**

In [60]:
import math
def phi_euler(n):
    contador = 0
    valores_phi = []
    for p in range(1, n+1):
        if math.gcd(p, n) == 1:
            valores_phi.append(p)
            contador+=1
    print("Coprimos:", valores_phi)
    return contador

**Vamos agora testar a nossa função com os exemplos dados $\varphi(7),\varphi(10)$:**

In [61]:
print(phi_euler(7))

Coprimos: [1, 2, 3, 4, 5, 6]
6


In [62]:
print(phi_euler(10))

Coprimos: [1, 3, 7, 9]
4


**Agora, com a função testada, vamos chamar a função phi de euler para $\varphi(35),\varphi(36),\varphi(37)$ e, logo em seguida, vamos chamar também a função geradores para cada respectivo $n$. Com isso, será possível identificar uma relação entre elas.**

In [63]:
print(phi_euler(35))
print("Geradores:",gerador(35))


Coprimos: [1, 2, 3, 4, 6, 8, 9, 11, 12, 13, 16, 17, 18, 19, 22, 23, 24, 26, 27, 29, 31, 32, 33, 34]
24
Geradores: [1, 2, 3, 4, 6, 8, 9, 11, 12, 13, 16, 17, 18, 19, 22, 23, 24, 26, 27, 29, 31, 32, 33, 34]


In [64]:
print(phi_euler(36))
print("Geradores:", gerador(36))

Coprimos: [1, 5, 7, 11, 13, 17, 19, 23, 25, 29, 31, 35]
12
Geradores: [1, 5, 7, 11, 13, 17, 19, 23, 25, 29, 31, 35]


In [65]:
print(phi_euler(37))
print("Geradores:", gerador(37))

Coprimos: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36]
36
Geradores: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36]


**Pode-se perceber que os termos geradores são exatamente os mesmos termos que compõem a função phi de Euler.**


A função phi de Euler possui muitas propriedades interessantes, incluindo:

1. **Multiplicativa**: Se $\text{mdc}(a, b) = 1$, então $\varphi(ab) = \varphi(a)\varphi(b)$.
2. **Teorema de Euler**: Se $\text{mdc}(a, n) = 1$, então $a^{\varphi(n)} \equiv 1 \mod n$.



### Grupos multiplicativos

Para multiplicação, o elemento neutro é sempre $1$, e a definição de um grupo requer que cada elemento tenha um inverso. É por isso que, módulo $N$, precisamos nos restringir aos elementos $a$ que são coprimos com $N$, pois, caso contrário, $a^{-1}$ não existe.

**<span style="color: red;">[4.3] Mostre que $5$ módulo $35$ não tem inverso multiplicativo. </span>**

**Existem duas formas de provar se um número tem inverso multiplicativo:**

>**Podemos testar todas as possiblididades, até encontrar uma em que $a \cdot b \equiv 1 \pmod n$, sendo $a$ o número testado e $b$ o inverso multiplicativo;**


In [70]:
def check_inverso_multiplicativo(a, amod):
    tem_inverso_multiplicativo = False
    for b in range(1, amod):
        aux = 0
        aux = a * b
        if aux % amod == 1:
            tem_inverso_multiplicativo = True
            print(b,"é inverso multiplicativo de", a, "em mod", amod)
            return
    print(a, "não tem inverso multplicativo em mod", amod)
    return  

In [71]:
check_inverso_multiplicativo(5, 35)

5 não tem inverso multplicativo em mod 35



>**Ou podemos somente verificar se $a$ e $n$ são coprimos, caso forem, existe um inverso multiplicativo para $a$ em $mod$ n**

In [72]:
def check_facil_inverso_multiplicativo(a, amod):
    if math.gcd(a, amod) == 1:
        print(a, "tem inverso multiplicativo em mod", amod)
        return
    print(a, "não tem inverso multiplicativo em mod", amod)

In [73]:
check_facil_inverso_multiplicativo(5, 35)

5 não tem inverso multiplicativo em mod 35


**Definição**
O **grupo multiplicativo módulo $N$**, denotado como $(\mathbb{Z}/N\mathbb{Z})^*$ ou $\mathbb{Z}_N^*$, consiste em todos os inteiros entre $1$ e $N-1$ que são **coprimos com $N$** (ou seja, $\text{mdc}(a, N) = 1$).

Este é um grupo sob multiplicação módulo $N$ e possui as seguintes propriedades:

1. **Fechamento**: Se $a, b \in \mathbb{Z}_N^*$, então $a \times b \mod N \in \mathbb{Z}_N^*$.
2. **Associatividade**: $(a \times b) \times c \equiv a \times (b \times c) \mod N$.
3. **Elemento neutro**: O número $1$ é a identidade multiplicativa/elemento neutro.
4. **Inversos**: Todo elemento $a$ possui um único inverso $a^{-1}$ tal que $a \times a^{-1} \equiv 1 \mod N$.
5. **Abeliano**: $a \times b \equiv b \times a \mod N$.
6. **Ordem**: O **tamanho** do grupo é dado pela função totiente de Euler $\phi(N)$.

##### Exemplo 1. Módulo $15$ ($\mathbb{Z}_{15}^*$)

O grupo multiplicativo $\mathbb{Z}_{15}^*$ consiste em inteiros módulo 15 que são coprimos com 15 (ou seja, elementos invertíveis sob multiplicação).

$$
\mathbb{Z}_{15}^* = \{1, 2, 4, 7, 8, 11, 13, 14\}
$$
Confirma-se: $|\mathbb{Z}_{15}^*|=\varphi(15) = (5-1)(3-1)= 4 \cdot 2 = 8$.

Eis a tabela de multiplicação do grupo $\mathbb{Z}_{15}^*$:
$$
\begin{array}{c|cccccccc}
\times \ (\text{mod } 15) & 1 & 2 & 4 & 7 & 8 & 11 & 13 & 14 \\
\hline
1 & 1 & 2 & 4 & 7 & 8 & 11 & 13 & 14 \\
2 & 2 & 4 & 8 & 14 & 1 & 7 & 11 & 13 \\
4 & 4 & 8 & 1 & 13 & 2 & 14 & 7 & 11 \\
7 & 7 & 14 & 13 & 4 & 11 & 2 & 1 & 8 \\
8 & 8 & 1 & 2 & 11 & 4 & 13 & 14 & 7 \\
11 & 11 & 7 & 14 & 2 & 13 & 1 & 8 & 4 \\
13 & 13 & 11 & 7 & 1 & 14 & 8 & 4 & 2 \\
14 & 14 & 13 & 11 & 8 & 7 & 4 & 2 & 1 \\
\end{array}
$$

**Observações Principais:**
1. **Elemento Identidade**: $1$ (já que $a \cdot 1 \equiv a \mod 15$).
2. **Inversos**:
- $1^{-1} = 1$
- $2^{-1} = 8$ (já que $2 \cdot 8 = 16 \equiv 1 \mod 15$)
- $4^{-1} = 4$
- $7^{-1} = 13$
- $14^{-1} = 14$ (autoinverso).
3. **Ordem dos Elementos**:
- $2$ tem ordem 4 ($2^4 = 16 \equiv 1 \mod 15$).
- $4$ tem ordem 2.
- $7, 13, 14$ têm ordem 2.
- $8, 11$ têm ordem 4.

**Estrutura do Grupo:**

Na Aula 3 vimos que $\mathbb{Z}_3 \times \mathbb{Z}_5 = \{(g,h) \mid 0 \leq g < 3 \wedge 0 \leq h < 5 \}$. Da mesma form temos que $\mathbb{Z}_3^* \times \mathbb{Z}_5^*$ é equivalente (exceto pela notação) a $\mathbb{Z}_{15}^*$:

$$
\mathbb{Z}_3^* \times \mathbb{Z}_5^* = \mathbb{Z}_{15}^*
$$

$$
\begin{array}{c|ccccc}
&0&1&2&3&4\\ \hline
0&-&-&-&-&- \\
1&-&1=(1,1)&7=(1,2)&13=(1,3)&4=(1,4) \\
2&-&11=(2,1)&2=(2,2)&8=(2,3)&14=(2,4) \\
\end{array}
$$

Esta tabela deixa claro que os inteiros $a$ tais que $\text{mdc}(a,15)=1$ não têm inverso.

##### Exemplo 2. Módulo $35$ ($\mathbb{Z}_{35}^*$)

- **Etapa 1: Fatorar $N$**
$35 = 5 \cdot 7$ (produto de dois primos distintos).

- **Etapa 2: Calcular $\phi(35)$**
Função totiente de Euler:
$
\phi(35) = \phi(5) \times \phi(7) = 4 \times 6 = 24.
$
Portanto, $\mathbb{Z}_{35}^*$ tem **24 elementos**.

- **Etapa 3: Listar os elementos coprimos de 35**
Todos os inteiros $1 \leq a < 35$ onde $\text{mdc}(a, 35) = 1$:
$
\mathbb{Z}_{35}^* = \{1, 2, 3, 4, 6, 8, 9, 11, 12, 13, 16, 17, 18, 19, 22, 23, 24, 26, 27, 29, 31, 32, 33, 34\}.
$

- **Etapa 4: Verificar os inversos**
Cada elemento tem um inverso único módulo 35. Por exemplo:
- $2^{-1} \mod 35 = 18$ (já que $2 \times 18 = 36 \equiv 1 \mod 35$).
- $6^{-1} \mod 35 = 6$ (já que $6 \cdot 6 = 36 \equiv 1 \mod 35$).

- **Etapa 5: Verifique a Estrutura Cíclica**
Como $35 = 5 \cdot 7$ (primos $\neq 2$), $\mathbb{Z}_{35}^*$ é **cíclico** (possui um gerador).
- **Exemplo de gerador**: $2$ (verifique se sua ordem é 24).

##### Exemplo 3. Módulo $143$ ($\mathbb{Z}_{143}^*$)

- **Etapa 1: Fatorar $N$**
$143 = 11 \cdot 13$.

- **Etapa 2: Calcular $\phi(143)$**
$
\phi(143) = \phi(11) x \phi(13) = 10 x 12 = 120.
$
Portanto, $\mathbb{Z}_{143}^*$ tem **120 elementos**.

- **Etapa 3: Listar os Elementos (Parcialmente)**
Elementos coprimos com 143 (muitos para listar completamente, mas exemplos incluem):
$
\mathbb{Z}_{143}^* = \{1, 2, 3, \dots, 142\} \setminus \{11, 13, 22, 26, \dots, 130, 143\}.
$
(Exclui múltiplos de 11 ou 13.)

- **Etapa 4: Verificar Inversos**
- $2^{-1} \mod 143 = 72$ (já que $2 \cdot 72 = 144 \equiv 1 \mod 143$).
- $7^{-1} \mod 143 = 41$ (já que $7 \cdot 41 = 287 \equiv 1 \mod 143$).

**<span style="color: red;">[4.5] Liste todos os elementos de* $\mathbb{Z}_{24}^*$ *e faça a uma tabela de multiplicação deste grupo.**

In [74]:
phi_euler(24)

Coprimos: [1, 5, 7, 11, 13, 17, 19, 23]


8

$$
\begin{array}{c|cccccccc}
&1&5&7&11&13&17&19&23\\ \hline
1&1&5&7&11&13&17&19&23 \\
5&5&1&11&7&17&13&23&19 \\
7&7&11&1&5&19&23&13&17 \\
11&11&7&5&1&23&19&17&13 \\
13&13&17&19&23&1&5&7&11 \\
17&17&13&23&19&5&1&11&7\\
19&19&23&13&17&7&11&1&5 \\
23&23&19&17&13&11&7&5&1 \\
\end{array}
$$

**<span style="color: red;">[4.6] Encontre todos os geradores de* $\mathbb{Z}_{35}^*$ *(Dica: Teste potências de 2 módulo 35.) -- ANULADO</span>**

**<span style="color: red;">[4.7] Calcule $5^{-1} \mod 143$.</span>**

In [75]:
check_inverso_multiplicativo(5, 143)

86 é inverso multiplicativo de 5 em mod 143


**Essa é uma das formas de encontrar o inverso multiplicativo de um número $a$ em módulo $N$. Mais adiante, vamos entender como fazer isso usando coeficientes e o Algoritmo Euclidiano Estendido.**

# Como Calcular Inversos Modulares Usando o Algoritmo Euclidiano Estendido

Calcular o inverso multiplicativo por tentativa e erro consome muito tempo para $N$ grandes. Felizmente, existe um algoritmo eficiente: calcular inversos é um caso particular do Algoritmo Euclidiano Estendido.

#### **Teorema**
Sejam dois inteiros $a$ e $b$ (ambos não nulos), e seja  $d=\text{mdc}(a, b)$ o máximo divisor comum de $a$ e $b$. Então existem inteiros $x$ e $y$ tais que:
$$
a x + b y = d
$$

Este teorema também é conhecido como Teorema de Bézout, e $x$ e $y$ são às vezes chamados de **coeficientes de Bézout**. Esses valores são calculados por uma versão estendida do algoritmo de Euclides.

Observações:

1. O $\text{mdc}$ de $a$ e $b$ pode sempre ser expresso como uma combinação linear inteira de $a$ e $b$.

2. A equação $a x + b y = c$ tem soluções inteiras $(x, y)$ **se e somente se** $\text{mdc}(a, b)$ divide $c$.
3. Os coeficientes $x$ e $y$ não são únicos (existem infinitas soluções).

**Exemplo** Considere o algoritmo normal apresentado na Aula 1 para $a=1057, b=315, \text{mdc}(a,b)=???$

Defina $a=r_0=1057; b=r_1=315$
$$
\begin{array}{ccccc|ll}
1057 &=& 3 \cdot 315 &+& 112 & q_1=3&r_2=112 \\
315 &=& 2 \cdot 112 &+& 91 & q_2=2&r_3=91 \\
112 &=& 1 \cdot 91 &+& 21 & q_3=1&r_4=21 \\
91 &=& 4 \cdot 21 &+& 7 & q_4=4&r_5=7 \\
21 &=& 3 \cdot 7 &+& 0 & q_5=3&r_6=0 \leftarrow \\
\end{array}
$$
Se $r_k=0$ então $r_{k-1}=\text{mdc}$

Compare a versão simples com a versão estendida. O quociente e o resto não mudam, mas o algoritmo estendido calcula mais valores intermediários em cada etapa. O algoritmo começa com duas equações padrão e, em seguida, continua subtraindo a última equação $q_i$ vezes da equação anterior:

$$
\begin{array}{rcrcc|c}
1 \cdot 1057 & + & 0 \cdot 315 & = & 1057 & \qquad q_i\\
0 \cdot 1057 & + & 1 \cdot 315 & = & 315 & \mathsf{subtrair}\ 3\ \mathsf{vezes} \\
1 \cdot 1057 & + & -3 \cdot 315 & = & 112 & \mathsf{subtrair}\ 2\ \mathsf{vezes} \\
-2 \cdot 1057 & + & 7 \cdot 315 & = & 91 & \mathsf{subtrair}\ 1\ \mathsf{vezes} \\
3 \cdot 1057 & + & -10 \cdot 315 & = & 21 & \mathsf{subtrair}\ 4\ \mathsf{vezes} \\
-14 \cdot 1057 & + & 47 \cdot 315 & = & 7 &\\
x \cdot a\quad\ \ &+&y \cdot b\quad&=&d
\end{array}
$$
Então, descobrimos que $x=-14, y=47$.

Verificando: $-14 \cdot 1057 + 47 \cdot 315 = -14\,798 + 14\,805 = 7$

**<span style="color: red;">[4.8] Use este algoritmo para calcular $x$ e $y$ para $a=23$ e $b=101$.**


**Queremos encontrar x e y, para os quais ax + by = d, ou seja, queremos encontrar os coeficientes de Bézout:**

In [76]:
def algoritmo_euclides_estendido(a, b):
    x_anterior, y_anterior, r_anterior = 1, 0, a
    x_atual, y_atual, r_atual = 0, 1, b

    while r_atual != 0:
        q = r_anterior // r_atual
        r_prox = r_anterior % r_atual
        x_prox = x_anterior - q * x_atual
        y_prox = y_anterior - q * y_atual
        
        print(f"Coeficientes de Bézout atuais: x = {x_atual} e y = {y_atual} e o resto é {r_atual}")

        x_anterior, y_anterior, r_anterior = x_atual, y_atual, r_atual
        x_atual, y_atual, r_atual = x_prox, y_prox, r_prox

    print(f"MDC({a}, {b}) = {r_anterior}")
    return r_anterior, x_anterior, y_anterior

**Vamos testar, para o exemplo, para garantir que a nossa função funciona:**

In [77]:
algoritmo_euclides_estendido(315, 1057)

Coeficientes de Bézout atuais: x = 0 e y = 1 e o resto é 1057
Coeficientes de Bézout atuais: x = 1 e y = 0 e o resto é 315
Coeficientes de Bézout atuais: x = -3 e y = 1 e o resto é 112
Coeficientes de Bézout atuais: x = 7 e y = -2 e o resto é 91
Coeficientes de Bézout atuais: x = -10 e y = 3 e o resto é 21
Coeficientes de Bézout atuais: x = 47 e y = -14 e o resto é 7
MDC(315, 1057) = 7


(7, 47, -14)

In [78]:
algoritmo_euclides_estendido(23, 101)

Coeficientes de Bézout atuais: x = 0 e y = 1 e o resto é 101
Coeficientes de Bézout atuais: x = 1 e y = 0 e o resto é 23
Coeficientes de Bézout atuais: x = -4 e y = 1 e o resto é 9
Coeficientes de Bézout atuais: x = 9 e y = -2 e o resto é 5
Coeficientes de Bézout atuais: x = -13 e y = 3 e o resto é 4
Coeficientes de Bézout atuais: x = 22 e y = -5 e o resto é 1
MDC(23, 101) = 1


(1, 22, -5)

**Nossa função retorna (d, coeficiente X, coeficiente Y), tal que $ax + by = d$.**

**<span style="color: red;">[4.9] Modifique sua implementação do algoritmo de Euclides na Parte 1 para calcular $x$ e $y$ (os coeficientes de Bézout). Use seu algoritmo para calcular $\text{mdc}(19,1001)$ usando este método.**

In [79]:
algoritmo_euclides_estendido(19, 1001)

Coeficientes de Bézout atuais: x = 0 e y = 1 e o resto é 1001
Coeficientes de Bézout atuais: x = 1 e y = 0 e o resto é 19
Coeficientes de Bézout atuais: x = -52 e y = 1 e o resto é 13
Coeficientes de Bézout atuais: x = 53 e y = -1 e o resto é 6
Coeficientes de Bézout atuais: x = -158 e y = 3 e o resto é 1
MDC(19, 1001) = 1


(1, -158, 3)

**Novamente, dentro dos parênteses temos (d, coeficiente X, coeficiente Y), tal que $ax + by = d$.**

### Calculando inversos

A AEE pode ser usada para calcular inversos multiplicativos de forma eficiente:

**Corolário**: Se $a$ e $N$ são primos entre si ($\text{mdc}(a, N) = 1$), então $a$ tem um inverso multiplicativo $a^{-1}$ módulo $N$.

O **Algoritmo Euclidiano Estendido (AEE)** resolve para inteiros $x$ e $y$ tais que:
$$
a \cdot x + N \cdot y = \text{mdc}(a, N)
$$
Quando $\text{mdc}(a, N) = 1$, a equação se torna:
$$
a \cdot x \equiv 1 \mod N
$$
Aqui, $x$ é o **inverso multiplicativo** de $a$ módulo $N$, denotado $a^{-1} \mod N$.

**<span style="color: red;">[4.10] Mostre como calcular $23^{-1} \mod 101$ usando o resultado da questão 4.8. Dica: considere $ax+by=d$ módulo $101$.</span>**

**Na questão 4.8, vimos que o $\text{mdc}(23, 101) = 1$, ou seja, são primos entre si. Pelo Corolário, se dois números são primos entre si, então $a$ tem um inverso multiplicativo $a^{-1}$ módulo $N$. Isso quer dizer que 23 possui inverso multiplicativo em módulo 101. Assim, vamos rodar novamente a função para encontrar o MDC e os coeficientes de Bézout. O coeficiente tem que ser o número k, tal que $23 \cdot k \equiv 1$ mod $101$.**

**Vamos também rodar a nossa função prévia, que encontra o inverso multiplicativo, para conferir o nosso resultado. Espera-se que o coeficiente X seja igual ao inverso multiplicativo encontrado pela função:**

In [80]:
print(algoritmo_euclides_estendido(23, 101))
check_inverso_multiplicativo(23, 101)

Coeficientes de Bézout atuais: x = 0 e y = 1 e o resto é 101
Coeficientes de Bézout atuais: x = 1 e y = 0 e o resto é 23
Coeficientes de Bézout atuais: x = -4 e y = 1 e o resto é 9
Coeficientes de Bézout atuais: x = 9 e y = -2 e o resto é 5
Coeficientes de Bézout atuais: x = -13 e y = 3 e o resto é 4
Coeficientes de Bézout atuais: x = 22 e y = -5 e o resto é 1
MDC(23, 101) = 1
(1, 22, -5)
22 é inverso multiplicativo de 23 em mod 101


**Como podemos ver, como é retornado pela primeira função, temos (d, coeficiente X, coeficiente Y), tal que $ax + by = d$. Assim, o coeficiente X é igual ao valor k encontrado pela outra função, responsável por encontrar um inverso multiplicativo, que é 22.**

**<span style="color: red;">[4.11] Use 4.9 para calcular $19^{-1} \mod 1001$. </span>**

**Agora que sabemos que para encontrar o inverso multiplicativo de $a^{-1}$ módulo $N$, basta encontrar o coeficiente X, tal que $ax + by = d$, com $a = 19$ e $N = 1001$  podemos rodar o nosso algoritmo para encontrar tal coeficiente:**

In [82]:
print(algoritmo_euclides_estendido(19, 1001))

Coeficientes de Bézout atuais: x = 0 e y = 1 e o resto é 1001
Coeficientes de Bézout atuais: x = 1 e y = 0 e o resto é 19
Coeficientes de Bézout atuais: x = -52 e y = 1 e o resto é 13
Coeficientes de Bézout atuais: x = 53 e y = -1 e o resto é 6
Coeficientes de Bézout atuais: x = -158 e y = 3 e o resto é 1
MDC(19, 1001) = 1
(1, -158, 3)


**Logo, podemos concluir que o inverso multiplicativo de $19$ em módulo $1001$ é $-158$, tal que $19 \cdot (-158) \equiv 1$ mod $1001$**